pkg/frontend/encryptionathost_validation.go

```diff
@@ -0,0 +1,71 @@
+package frontend
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/resources/armfeatures"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient/mgmt/features"
+)
+
+type EncryptionAtHostValidator interface {
+	ValidateEncryptionAtHost(ctx context.Context, subscriptionID string, oc *api.OpenShiftCluster) error
+}
+
+type encryptionAtHostValidator struct{}
+
+func (e encryptionAtHostValidator) ValidateEncryptionAtHost(ctx context.Context, subscriptionID string, oc *api.OpenShiftCluster) error {
+	credential, err := azidentity.NewDefaultAzureCredential(nil)
+	if err != nil {
+		return err
+	}
+
+	subFeatureRegistrationsClient, err := armfeatures.NewSubscriptionFeatureRegistrationsClient(subscriptionID, credential, nil)
+	if err != nil {
+		return err
+	}
+	return validateEncryptionAtHost(ctx, subFeatureRegistrationsClient, oc)
+}
+
+func validateEncryptionAtHost(ctx context.Context, subFeatureRegistrationsClient features.SubscriptionFeatureRegistrationsClient, oc *api.OpenShiftCluster) error {
+	var hasEncryptionAtHostEnabled bool
+	profilesToCheck := append([]api.WorkerProfile{{EncryptionAtHost: oc.Properties.MasterProfile.EncryptionAtHost}}, oc.Properties.WorkerProfiles...)
+	for _, profile := range profilesToCheck {
+		if profile.EncryptionAtHost == api.EncryptionAtHostEnabled {
+			hasEncryptionAtHostEnabled = true
+			break
+		}
+	}
+	if hasEncryptionAtHostEnabled {
+		err := IsRegisteredForEncryptionAtHostFeature(ctx, subFeatureRegistrationsClient)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func IsRegisteredForEncryptionAtHostFeature(ctx context.Context, subFeatureRegistrationsClient features.SubscriptionFeatureRegistrationsClient) error {
+	response, err := subFeatureRegistrationsClient.Get(ctx, "Microsoft.Compute", "EncryptionAtHost", nil)
+	if err != nil {
+		return err
+	}
+	if *response.Properties.State == armfeatures.SubscriptionFeatureRegistrationStateRegistered {
+		return nil
+	}
+	return &api.CloudError{
```

Comment:
This is kind of a nit, but I think it reads better to `return nil` as default, flip the conditional on line 60 to `if *response.Properties.State != armfeatures.SubscriptionFeatureRegistrationStateRegistered`, and then return the cloud error there instead.

Reply:
@cadenmarchese This was a good call out. Made changes as mentioned.